### PR TITLE
Task 2

### PR DESCRIPTION
**Pull request by virtushawk**
Pull request for module 2

### CODE REVIEW

File: controller/src/main/java/com/epam/esm/controller/TagController.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    /**
+     * Gets tag.
+     *
+     * @param id the id
+     * @return the tag
+     */
+    @GetMapping(value = "/{id}")
+    @ResponseStatus(HttpStatus.CREATED)
```
Comment (SenPavel) [inline controller/src/main/java/com/epam/esm/controller/TagController.java]: The "get" method shouldn't return 201.

File: controller/src/main/java/com/epam/esm/controller/TagController.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    /**
+     * Create tag
+     *
+     * @param tagDTO        the tag dto
+     * @param bindingResult the binding result
+     * @return the tag dto
+     */
+    @PostMapping
```
Comment (SenPavel) [inline controller/src/main/java/com/epam/esm/controller/TagController.java:51]: Please provide correct response status as PostMapping returns 200 by default

File: controller/src/main/java/com/epam/esm/controller/CustomExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @param exception the exception
+     * @param locale    the locale
+     * @return the response entity
+     */
+    @ExceptionHandler(CertificateNotFoundException.class)
+    public ResponseEntity<ErrorResponse> handleControllerNotFoundException(CertificateNotFoundException exception, Locale locale) {
+        String errorMessage = messageSource.getMessage("error.certificateNotFound", new Object[]{}, locale);
+        ErrorResponse response = new ErrorResponse();
+        response.setErrorMessage(errorMessage + ' ' + exception.getMessage());
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CustomExceptionHandler.java]: Please review this class and decide if any relevant constants should be declared.

File: controller/build.gradle [hunk 4]
```diff
@@ -0,0 +1,14 @@
+plugins {
+    id 'war'
+    id 'java'
```
Comment (maximelyashevich) [inline controller/build.gradle]: Do we need `java` plugin for this module? What's about `allprojects` configuration? Please review other modules as well.

File: service/src/main/java/com/epam/esm/dto/CertificateDTO.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package com.epam.esm.dto;
+
+import javax.validation.constraints.*;
```
Comment (maximelyashevich) [inline service/src/main/java/com/epam/esm/dto/CertificateDTO.java]: Please replace it with single import statements.

File: service/src/test/java/com/epam/esm/service/CertificateServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    @Mock
+    private ObjectMapper objectMapper;
+
+    @Test
+    void createValid() {
+        Certificate certificate = Certificate.builder().name("test name").description("test description").price(new BigDecimal("10"))
+                .duration(5).tags(new HashSet<>()).build();
+        certificate.getTags().add(Tag.builder().name("test tag").build());
+        CertificateDTO expected = CertificateDTO.builder().name("test name").description("test description").price(new BigDecimal("10"))
+                .duration(5).tags(new HashSet<>()).build();
```
Comment (SenPavel) [inline service/src/test/java/com/epam/esm/service/CertificateServiceImplTest.java]: Please move the object creation/params setting to @Before/BeforeEach method as these test methods are difficult to read and understand what is being tested.

**Review by SenPavel — CHANGES_REQUESTED** [review]

File: controller/src/main/java/com/epam/esm/controller/CustomExceptionHandler.java (hunk 3 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/CustomExceptionHandler.java]: done!

File: controller/build.gradle (hunk 4 above)
Comment (virtushawk) [reply controller/build.gradle]: done!

File: service/src/main/java/com/epam/esm/dto/CertificateDTO.java (hunk 5 above)
Comment (virtushawk) [reply service/src/main/java/com/epam/esm/dto/CertificateDTO.java]: done!

File: controller/src/main/java/com/epam/esm/controller/TagController.java (hunk 1 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/TagController.java]: done!

File: controller/src/main/java/com/epam/esm/controller/TagController.java (hunk 2 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/TagController.java:51]: done!

**Review by SenPavel — APPROVED** [review]

**Review by ykazlovich — APPROVED** [review]

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side.

Regards,
Maks

**Pull request merged**